airflow/migrations/versions/83f031fd9f1c_change_ts_columns_to_datetime_on_mssql.py

```diff
@@ -0,0 +1,187 @@
+    """
+    for constraint, columns in constraint_dict.items():
+        if column_name in columns:
+            if constraint[1].lower().startswith("primary"):
+                operator.create_primary_key(constraint_name=constraint[0], columns=columns)
+            elif constraint[1].lower().startswith("unique"):
+                operator.create_unique_constraint(constraint_name=constraint[0], columns=columns)
+
+
+def __use_date_time2(conn):
+    result = conn.execute(
+        """SELECT CASE WHEN CONVERT(VARCHAR(128), SERVERPROPERTY ('productversion'))
+        like '8%' THEN '2000' WHEN CONVERT(VARCHAR(128), SERVERPROPERTY ('productversion'))
+        like '9%' THEN '2005' ELSE '2005Plus' END AS MajorVersion"""
+    ).fetchone()
+    mssql_version = result[0]
+    return mssql_version not in ("2000", "2005")
+
+
+def __is_timestamp(conn, table_name, column_name):
+    query = """SELECT
+    TYPE_NAME(C.USER_TYPE_ID) AS DATA_TYPE
+    FROM SYS.COLUMNS C
+    JOIN SYS.TYPES T
+    ON C.USER_TYPE_ID=T.USER_TYPE_ID
+    WHERE C.OBJECT_ID=OBJECT_ID('{table_name}') and C.NAME='{column_name}';
+    """.format(
+        table_name=table_name, column_name=column_name
+    )
+    column_type = conn.execute(query).fetchone()[0]
+    return column_type == "timestamp"
+
+
+def recreate_mssql_ts_column(conn, op, table_name, column_name):
+    """
+    Drop the timestamp column and recreate it as
+    datetime or datetime2(6)
+    """
+    if __is_timestamp(conn, table_name, column_name) and is_table_empty(conn, table_name):
+        with op.batch_alter_table(table_name) as batch_op:
+            constraint_dict = get_table_constraints(conn, table_name)
+            drop_column_constraints(batch_op, column_name, constraint_dict)
+            batch_op.drop_column(column_name=column_name)
+            if __use_date_time2(conn):
+                batch_op.add_column(sa.Column(column_name, mssql.DATETIME2(precision=6), nullable=False))
+            else:
+                batch_op.add_column(sa.Column(column_name, mssql.DATETIME, nullable=False))
+            create_constraints(batch_op, column_name, constraint_dict)
+
+
+def alter_mssql_datetime_column(conn, op, table_name, column_name, nullable):
+    """Update the datetime column to datetime2(6)"""
+    if __use_date_time2(conn):
+        op.alter_column(
+            table_name=table_name,
+            column_name=column_name,
+            type_=mssql.DATETIME2(precision=6),
+            nullable=nullable,
+        )
+
+
+def alter_mssql_datetime2_column(conn, op, table_name, column_name, nullable):
+    """Update the datetime2(6) column to datetime"""
+    if __use_date_time2(conn):
+        op.alter_column(
+            table_name=table_name, column_name=column_name, type_=mssql.DATETIME, nullable=nullable
+        )
+
+
+def upgrade():
+    """Change timestamp and datetime to datetime2/datetime when using MSSQL as backend"""
```

Comment:
Can you explain to me what the right column type is to use when?

Reply:
Both are timezone-less datetime types. `datetime2` is an improved version of `datetime` and stores a strict superset of values, and should always be preferred if possible. According to `__use_date_time2()` (implemented in this PR), `datetime2` is chosen by default, and `dateimt` only on MSSQL 2000 and 2005 (persuambly because `datetime2` isn’t available on those versions, I did not check).

One semi-related issue here is, how does Airflow (and potentially SQLAlchemy) handle timezones? MSSQL’s datetime/2 types do not contain timezone information, and does not perform timezone conversion on `select`. But MySQL’s `timestamp` automatically performs timezone conversion `select` (it always stores values in UTC). So there may be subtle issues if this is not handled with care.

Reply:
In Airflow we try pretty hard to only ever store UTC in the DB - see https://github.com/apache/airflow/blob/2a2adb3f94cc165014d746102e12f9620f271391/airflow/utils/sqlalchemy.py#L41-L91

Reply:
Cool, in that case the column type conversion should be correct IMO.